content/blog/2022/2022-03-31-be-informed-with-lombok.md

```diff
@@ -0,0 +1,474 @@
+The Repository class that fetches data from the table is as below:
+
+````java
+    @Repository
+    public interface BookRepository extends JpaRepository<Book, Long> {
+    }
+````
+There are **three** main problems here:
+1. In an entity class, not all attributes of an entity are initialized. If an attribute has a FetchType of association LAZY,
+it gets invoked only when used in the application. However, **@ToString would compute all attributes 
+of an entity(irrespective of whether it is used) making one or multiple database calls** which can unintentionally **cause performance issues**.
+2. Further, if we **call toString() on the entity outside the Transaction scope**, it could lead to `LazyInitializationException`.
+3. In the case of associations like @ManyToMany between 2 entities, **logging the entity data could result in evaluating circular references and causing StackOverflowError**. In the example above, the `Book` entity 
+will try to fetch all authors of the book. The `Author` entity in turn will try to find all books of the author. This process will keep repeating until it results in an error.
+
+**2. `Avoid @EqualsAndHashCode :`**
+Lombok uses all non-final attributes to evaluate and override default equals and hashCode. This isn't always desirable in the case of entities due to the following reasons:
+1. Most primary keys in the database are **auto-generated** (either using sequences or UUID) 
+during insertion. This **can cause issues in the hashCode computation process**
+as the `ID` is not available beforehand causing unexpected results.
+2. **Every database record is uniquely identified by its primary key**. In such cases using the Lombok implementation of `@EqualsAndHashCode` might not be ideal. 
+
+Although Lombok allows us to include and exclude attributes, for the sake of brevity it might be a **better option to 
+override these methods (toString(), equals(), hashcode()) ourselves** and not rely on Lombok.
+
+### Lombok Annotations hide violations
+Consider a model class in the example below:
+ ````java
+    @Data
+    @Builder
+    @AllArgsConstructor
+    public class CustomerDetails {
+    
+        private String id;
+        private String name;
+        private String buildingNm;
+        private String blockNo;
+        private String streetNm;
+        private String city;
+        private int postcode;
+        private String state;
+        private String country;
+        private Gender gender;
+        private String dateOfBirth;
+        private String email;
+        private String phoneNo;
+        private String drivingLicenseNo;
+        private String licenseIssueState;
+    }
+ ````
+The above class has the following **flaws**:
+1. It is a **poorly designed model class**. We could consider restructuring the fields to create separate classes to store Address and LicenseDetails.
+2. **@Builder hides the complexity of autowiring instances** especially if custom objects 
```

Comment:
I'm not sure I understand this point. What is the "complexity of autowiring instances"? Can you elaborate? How does it tie into this section's title about "hiding validations"?

Reply:
@thombergs  I mainly want to highlight 4 points
1. Sonar does not catch the violations when annotations are added.
2. A class having a huge constructor means something is wrong. Maybe the class has too much responsibility.
3. Using the right combination of @Builder and @RequiredArgsConstructor is essential so that the object created is not in an invalid state.
4. If a constructor has too many same type params. eg. too many String params, it is easy to define the object where the params can get out of order which might be difficult to trace.

Need some help structuring here. I am not sure of each of these points should be a separate heading or should i combine some of them together.

Reply:
Also will try to restructure the validation better by showing before Lombok result and after Lombok result.

Reply:
I suggest to structure it like this:

* section "Lombok Hides Coding Violations"
  * explains checkstyle etc. and how Lombok hides their violations
* section "`@AllArgsConstructor` May Introduce Errors When Refactoring"
  * explains that the order of constructor parameters changes when you change the order of fields
* section "`@Builder` Allows Creation of Invalid Objects
  * explains how `@Builder` lets us create objects in invalid states
 
each section with its own small code example and explanation.